File: Amazon Locker System/Java/AmazonLocker.java
package amazon_locker;

import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.locks.*;

// --- Core Domain ---

enum Size {
    SMALL, MEDIUM, LARGE
}

class Ticket {
    final String lockerId;
    final String code;
    final long creationTime;

    public Ticket(String lockerId, String code, long creationTime) {
        this.lockerId = lockerId;
        this.code = code;
        this.creationTime = creationTime;
    }

    @Override
    public String toString() {
        return "Ticket{lockerId='" + lockerId + "', code='" + code + "', creationTime=" + creationTime + "}";
    }
}

// --- Infrastructure / Interfaces ---

interface Clock {
    long currentTimeMillis();
}

class SystemClock implements Clock {
    @Override
    public long currentTimeMillis() {
        return System.currentTimeMillis();
    }
}

// --- Service ---

class LockerService {
    private final Clock clock;
    
    // 3 Queues for O(1) scaling logic
    private final EnumMap<Size, Deque<String>> queues = new EnumMap<>(Size.class);
    
    // O(1) Lookup for validation
    private final Map<String, Ticket> activeTickets = new ConcurrentHashMap<>();
    
    private final Lock lock = new ReentrantLock();

    public LockerService(Clock clock) {
        this.clock = clock;
        // Initialize queues
        queues.put(Size.SMALL, new ArrayDeque<>());
        queues.put(Size.MEDIUM, new ArrayDeque<>());
        queues.put(Size.LARGE, new ArrayDeque<>());
        
        // Fill Lockers (Simulation)
        for (int i = 0; i < 10; i++) queues.get(Size.SMALL).add("S-" + i);
        for (int i = 0; i < 10; i++) queues.get(Size.MEDIUM).add("M-" + i);
        for (int i = 0; i < 10; i++) queues.get(Size.LARGE).add("L-" + i);
    }

    /**
     * Input: size
     * Output: Ticket
     */
    public Ticket deposit(Size requestedSize) {
        lock.lock();
        try {
            String bestLockerId = findBestFit(requestedSize);
            if (bestLockerId == null) {
                throw new IllegalStateException("No locker available for size " + requestedSize);
            }

            // Generate simple logic for code
            String code = bestLockerId + "-" + (clock.currentTimeMillis() % 10000);
            
            Ticket ticket = new Ticket(bestLockerId, code, clock.currentTimeMillis());
            activeTickets.put(code, ticket);
            return ticket;
        } finally {
            lock.unlock();
        }
    }

    /**
     * Input: Locker Code
     * Output: Success message
     */
    public String pickup(String code) {
        // Validation is O(1)
        Ticket ticket = activeTickets.remove(code);
        if (ticket == null) {
            throw new IllegalArgumentException("Invalid or expired code: " + code);
        }

        lock.lock();
        try {
            returnLocker(ticket.lockerId);
        } finally {
            lock.unlock();
        }
        
        return "Locker " + ticket.lockerId + " opened. Package retrieved.";
    }

    /**
     * Removes all expired packages (older than 3 days)
     */
    public void cleanup() {
        long threeDaysInMillis = 3L * 24 * 60 * 60 * 1000;
        long now = clock.currentTimeMillis();
        
        lock.lock();
        try {
            Iterator<Map.Entry<String, Ticket>> iterator = activeTickets.entrySet().iterator();
            while (iterator.hasNext()) {
                Map.Entry<String, Ticket> entry = iterator.next();
                if (now - entry.getValue().creationTime > threeDaysInMillis) {
                    Ticket ticket = entry.getValue();
                    iterator.remove(); // Remove from active tickets
                    returnLocker(ticket.lockerId); // Return to pool
                    System.out.println("Expired package removed from " + ticket.lockerId);
                }
            }
        } finally {
            lock.unlock();
        }
    }

    // --- Helper Functions ---

    private String findBestFit(Size size) {
        // Cascading check: Requested -> Next Size Up
        switch (size) {
            case SMALL:
                String s = queues.get(Size.SMALL).poll();
                if (s != null) return s;
                // Fallthrough to Medium check if Small unavailable? 
                // In Java switch fallthrough logic is tricky if sharing return logic.
                // Explicit checks are cleaner.
                String m = queues.get(Size.MEDIUM).poll();
                if (m != null) return m;
                return queues.get(Size.LARGE).poll();
            case MEDIUM:
                String med = queues.get(Size.MEDIUM).poll();
                if (med != null) return med;
                return queues.get(Size.LARGE).poll();
            case LARGE:
                return queues.get(Size.LARGE).poll();
            default:
                return null;
        }
    }

    private void returnLocker(String lockerId) {
        if (lockerId.startsWith("S-")) queues.get(Size.SMALL).add(lockerId);
        else if (lockerId.startsWith("M-")) queues.get(Size.MEDIUM).add(lockerId);
        else if (lockerId.startsWith("L-")) queues.get(Size.LARGE).add(lockerId);
    }
}

// --- Driver / Test ---

class MockClock implements Clock {
    long time = 1000L;
    @Override
    public long currentTimeMillis() { return time; }
}

public class AmazonLocker {
    public static void main(String[] args) {
        System.out.println("=== Amazon Locker System Demo (Java) ===");
        
        MockClock mockClock = new MockClock();
        LockerService service = new LockerService(mockClock);

        // 1. DEPOSIT
        System.out.println("\n[Action] Deposit Small Package");
        Ticket ticket1 = service.deposit(Size.SMALL);
        System.out.println("Received Ticket: " + ticket1);

        // 2. DEPOSIT Next Size Up Logic
        // Drain smalls...
        for (int i = 0; i < 9; i++) service.deposit(Size.SMALL);
        
        System.out.println("\n[Action] Deposit Small Package (Small queues empty, expect Medium)");
        Ticket ticket2 = service.deposit(Size.SMALL);
        System.out.println("Received Ticket: " + ticket2 + " (Should be M-...)");

        // 3. PICKUP
        System.out.println("\n[Action] Pickup Ticket 1");
        System.out.println(service.pickup(ticket1.code));

        // 4. CLEANUP (Expired)
        System.out.println("\n[Action] Simulate 4 days pass...");
        Ticket large = service.deposit(Size.LARGE);
        System.out.println("Deposited Large package (to be expired): " + large);
        
        mockClock.time += 4L * 24 * 60 * 60 * 1000;
        
        System.out.println("[Action] Running Cleanup...");
        service.cleanup();
        
        System.out.println("\nDemo Completed.");
    }
}

File: Amazon Locker System/AmazonLocker.kt
package amazon_locker

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock
import java.util.ArrayDeque

// --- Core Domain ---

enum class Size {
    SMALL, MEDIUM, LARGE
}

data class Ticket(
    val lockerId: String,
    val code: String,
    val creationTime: Long
)

// --- Infrastructure / Interfaces ---

interface Clock {
    fun currentTimeMillis(): Long
}

class SystemClock : Clock {
    override fun currentTimeMillis() = System.currentTimeMillis()
}

// --- Service ---

class LockerService(private val clock: Clock) {
    // 3 Queues for O(1) scaling logic
    private val smallLockers = ArrayDeque<String>()
    private val mediumLockers = ArrayDeque<String>()
    private val largeLockers = ArrayDeque<String>()

    // O(1) Lookup for validation
    private val activeTickets = ConcurrentHashMap<String, Ticket>()

    private val lock = ReentrantLock()

    init {
        // Initialize lockers (Simulation)
        repeat(10) { smallLockers.add("S-$it") }
        repeat(10) { mediumLockers.add("M-$it") }
        repeat(10) { largeLockers.add("L-$it") }
    }

    /**
     * Input: size
     * Output: Locker Code (Tiket)
     */
    fun deposit(requestedSize: Size): Ticket {
        lock.withLock {
            val bestLockerId = findBestFit(requestedSize) 
                ?: throw IllegalStateException("No locker available for size $requestedSize")

            // Generate simple logic for code (UUID or random in real life)
            val code = "${bestLockerId}-${clock.currentTimeMillis() % 10000}"
            
            val ticket = Ticket(
                lockerId = bestLockerId,
                code = code,
                creationTime = clock.currentTimeMillis()
            )
            
            activeTickets[code] = ticket
            return ticket
        }
    }

    /**
     * Input: Locker Code
     * Output: Success message (and unlocking side effect)
     */
    fun pickup(code: String): String {
        // Validation is O(1)
        val ticket = activeTickets.remove(code) 
            ?: throw IllegalArgumentException("Invalid or expired code: $code")

        // Return locker to the correct queue
        lock.withLock {
            returnLocker(ticket.lockerId)
        }
        
        return "Locker ${ticket.lockerId} opened. Package retrieved."
    }

    /**
     * Removes all expired packages (older than 3 days)
     */
    fun cleanup() {
        val threeDaysInMillis = 3L * 24 * 60 * 60 * 1000
        val now = clock.currentTimeMillis()
        
        lock.withLock {
            // Iterate safely over a concurrent map snapshot or handle removing efficiently
            // Since we need to return lockers to queue, we need lock.
            // Note: In a real DB, this would be a query. Here strict checking.
            val iterator = activeTickets.entries.iterator()
            while (iterator.hasNext()) {
                val entry = iterator.next()
                if (now - entry.value.creationTime > threeDaysInMillis) {
                    val ticket = entry.value
                    iterator.remove() // Remove from active tickets
                    returnLocker(ticket.lockerId) // Return to pool
                    println("Expired package removed from ${ticket.lockerId}")
                }
            }
        }
    }

    // --- Helper Functions (Extensible) ---
    
    private fun findBestFit(size: Size): String? {
        // Cascading check: Requested -> Next Size Up
        return when (size) {
            Size.SMALL -> smallLockers.poll() ?: mediumLockers.poll() ?: largeLockers.poll()
            Size.MEDIUM -> mediumLockers.poll() ?: largeLockers.poll()
            Size.LARGE -> largeLockers.poll()
        }
    }

    private fun returnLocker(lockerId: String) {
        when {
            lockerId.startsWith("S-") -> smallLockers.add(lockerId)
            lockerId.startsWith("M-") -> mediumLockers.add(lockerId)
            lockerId.startsWith("L-") -> largeLockers.add(lockerId)
        }
    }
}

// --- Driver / Test ---

class MockClock : Clock {
    var time = 1000L
    override fun currentTimeMillis() = time
}

fun main() {
    println("=== Amazon Locker System Demo ===")
    
    val mockClock = MockClock()
    val service = LockerService(mockClock)

    // 1. DEPOSIT
    println("\n[Action] Deposit Small Package")
    val ticket1 = service.deposit(Size.SMALL)
    println("Received Ticket: $ticket1")

    // 2. DEPOSIT Next Size Up Logic (Consume all smalls first to test)
    // Draining smalls...
    repeat(9) { service.deposit(Size.SMALL) }
    
    println("\n[Action] Deposit Small Package (Small queues empty, expect Medium)")
    val ticket2 = service.deposit(Size.SMALL)
    println("Received Ticket: $ticket2 (Should be M-...)")

    // 3. PICKUP
    println("\n[Action] Pickup Ticket 1")
    println(service.pickup(ticket1.code))

    // 4. CLEANUP (Expired)
    println("\n[Action] Simulate 4 days pass...")
    service.deposit(Size.LARGE).also { println("Deposited Large package (to be expired): $it") }
    
    // Fast forward time
    mockClock.time += 4L * 24 * 60 * 60 * 1000 
    
    println("[Action] Running Cleanup...")
    service.cleanup()
    
    println("\nDemo Completed.")
}
